storage/src/main/java/tech/pegasys/artemis/storage/ChainStorageClient.java

```diff
@@ -306,4 +323,37 @@ public void onNewSlot(SlotEvent slotEvent) {
       }
     }
   }
+
+  private void updateCanonicalBlockIndex(final Bytes32 bestBlockRoot, final UnsignedLong slot) {
+    final NavigableMap<UnsignedLong, Bytes32> updatedIndices = new TreeMap<>();
+
+    UnsignedLong currentSlot = slot;
+    Bytes32 currentRoot = bestBlockRoot;
+    while (!Objects.equals(slotToCanonicalBlockRoot.get(currentSlot), currentRoot)) {
+      updatedIndices.put(currentSlot, currentRoot);
+      if (currentSlot.compareTo(UnsignedLong.valueOf(Constants.GENESIS_SLOT)) <= 0) {
+        // We've reached the genesis slot, nothing left to index
+        break;
+      }
+      currentRoot = store.getBlock(currentRoot).getParent_root();
+      BeaconBlock currentBlock = store.getBlock(currentRoot);
+      currentSlot = currentBlock.getSlot();
```

Comment:
The in-memory store of blocks doesn't go all the way back to genesis so at some point we'll get a `null` block from the store.  We could just stop iterating at that point as we'd have indexed every block still available in the store.

Reply:
Should we not provide the ability for the store to look up from disk if it doesn't have the data in memory?

Reply:
We will need to be able to lookup old blocks from disk by slot but the architecture needs working on.  I think the current setup is that the disk based storage is a separate "service" and so is used only via the event bus - that makes a lot of sense as you could potentially have one archive service storing data for multiple beacon chain or validator clients.

Older blocks would also be finalised so don't need to deal with re-orgs which may affect how we choose to store things (we may just store by slot number instead of hash which would allow range queries to be efficient).